src/meta/src/stream/stream_graph/fragment.rs

```diff
@@ -138,6 +143,28 @@ impl BuildingFragment {
 
         has_table
     }
+
+    /// Extract the required columns (in IDs) of each upstream table.
+    fn extract_upstream_table_columns(
+        // TODO: no need to take `&mut` here
+        fragment: &mut StreamFragment,
+    ) -> HashMap<TableId, Vec<i32>> {
+        let mut table_columns = HashMap::new();
+
+        visit::visit_fragment(fragment, |node_body| {
+            if let NodeBody::Chain(chain_node) = node_body {
+                let table_id = chain_node.table_id.into();
+                let column_ids = chain_node.upstream_column_ids.clone();
+                table_columns
+                    .try_insert(table_id, column_ids)
+                    .expect("currently there should be no two same upstream tables in a fragment");
```

Comment:
L160 asserts there should be no two same upstream tables in a fragment, but is it possible to have two different upstream tables?

Reply:
I think it's currently impossible. The reason for why using a collection here is explained in https://github.com/risingwavelabs/risingwave/pull/4740#discussion_r949840701.

Reply:
Thanks for your explanation.